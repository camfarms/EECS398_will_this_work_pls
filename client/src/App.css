.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}

div, button {
  margin: 10px;
}

div.MuiCollapse-wrapper, div.MuiCollapse-wrapperInner, div.MuiExpansionPanelDetails-root,
div.MuiPaper-root, div.MuiPaper-elevation1 div.makeStyles-root-137 div.MuiPaper-rounded,
div.makeStyles-tableWrapper-139 {
  margin: 0px;
  padding: 0px;
}

/* button {
  background: green;
  border-radius: 100px;
  color: white;
} */