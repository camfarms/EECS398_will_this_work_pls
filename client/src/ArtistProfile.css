div.artist {
    margin: 0;
    height: 200px;
    overflow: scroll;
}