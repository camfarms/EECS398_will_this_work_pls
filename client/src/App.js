import React, { Component } from 'react';
import './App.css';
import Spotify from 'spotify-web-api-js';
import NavBar from './components/headerComponent/navBar.js';
import Button from '@material-ui/core/Button';
import ButtonGroup from '@material-ui/core/ButtonGroup';
import ArtistProfile from './ArtistProfile.js';
import Lyrics from './Lyrics.js';
import 'bootstrap/dist/css/bootstrap.min.css';
import RecommendationsTable from './Recommendations/Recommendations.js';
import {createMuiTheme} from '@material-ui/core/styles';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles'
import { grey } from '@material-ui/core/colors';
import { CssBaseline } from '@material-ui/core';
import Grid from '@material-ui/core/Grid';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import Vibrant from 'node-vibrant';
import Hotkeys from 'react-hot-keys';
import Icon from '@material-ui/core/Icon';

const spotifyWebApi = new Spotify()
var timeRemaining = undefined;
var imageUrl = '';
var adaptive = false;
var themeMode = "dark";
var primary = '#4caf50';
var secondary = grey;
var play = true;
var shuffle = false;
var playIcon = <Icon>pause_circle_outline</Icon>;
var shuffleIcon = <Icon>shuffle</Icon>;

function componentToHex(c) {
  var hex = c.toString(16);
  return hex.length === 1 ? "0" + hex : hex;
}

function rgbToHex(r, g, b) {
  return "#" + componentToHex(r) + componentToHex(g) + componentToHex(b);
}

var theme = createMuiTheme( {
  palette: {
    type: themeMode,
    primary: {
      main: primary
    },
    secondary: secondary,
  },
});

class App extends Component {
  constructor() {
    super();
    const params = this.getHashParams();
    this.state = {
      loggedIn: params.access_token? true: false,
      nowPlaying: {
        name: 'Not Checked',
        artistName: 'Not Checked',
        songInfo: '',
        image: ''
      }
    }
    if (params.access_token) {
      spotifyWebApi.setAccessToken(params.access_token);
    }
  }

  getHashParams() {
    var hashParams = {};
    var e, r = /([^&;=]+)=?([^&;]*)/g,
        q = window.location.hash.substring(1);
    while ( e = r.exec(q)) {
       hashParams[e[1]] = decodeURIComponent(e[2]);
    }
    return hashParams;
  }
    
  getNowPlaying() {
    if (!(ArtistProfile === undefined)) {
      this.ArtistProfile.refreshArtist();
    }
    if (!(Lyrics === undefined)) {
      this.Lyrics.refreshArtist();
    }

    spotifyWebApi.getMyCurrentPlaybackState()
    .then((response) => {
      var tempName;
      var tempArtist;
      var tempImage;
      var songInfo;
      if (response.item === undefined) {
        tempName = 'You are not currently playing any songs on Spotify';
        tempArtist = 'You are not currently playing any songs on Spotify';
        songInfo = '';
        tempImage = undefined;
      }
      else {
        tempName = response.item.name;
        tempArtist = response.item.artists[0].name;
        tempImage = response.item.album.images[0].url;
        imageUrl = tempImage;
        songInfo = tempName + " - " + tempArtist;
        var songProgress = response.progress_ms;
        var songDuration = response.item.duration_ms;
        timeRemaining = songDuration - songProgress;
      }
      this.setState( {
        nowPlaying: {
          name: tempName,
          artistName: tempArtist,
          image: tempImage,
          songInfo: songInfo
        }
      })
    })
  }

  getNextSong() {
    spotifyWebApi.skipToNext();
    this.refresh();
    console.log(timeRemaining);
  }

  getLastSong() {
    spotifyWebApi.skipToPrevious();
    this.refresh();
    console.log(timeRemaining);
  }

  getPause(){    
    if (play) {
      spotifyWebApi.pause();
      playIcon = <Icon>play_circle_outline</Icon>;
      this.refresh();
      play = false;
    }
    else {
      spotifyWebApi.play();
      playIcon = <Icon>pause_circle_outline</Icon>;
      this.refresh();
      play = true;
    }
  }

  getShuffle(){
    if (!shuffle) {
      spotifyWebApi.setShuffle(true);
      shuffleIcon = <Icon>shuffle</Icon>;
      shuffle = true;
    }
    else {
      spotifyWebApi.setShuffle(false);
      shuffleIcon = <Icon>shuffle</Icon>;
      shuffle = false;
    }
  }

  // to get currently playing song on load
  componentDidMount() {
    this.getNowPlaying();
  }

  refresh() {
    this.getNowPlaying();
    if (adaptive) {
      this.updateTheme();
    }
    var self = this
    setTimeout(function() {
      self.getNowPlaying();
      if (adaptive) {
        self.updateTheme();
      }
    }, 250);
    clearTimeout();
    setTimeout(function() {
      self.getNowPlaying();
      if (adaptive) {
        self.updateTheme();
      }
    }, 500);
    clearTimeout();
  }

  defaultTheme() {
    adaptive = false;
    primary = '#4caf50';
    theme = createMuiTheme( {
      palette: {
        type: themeMode,
        primary: {
          main: primary
        },
        secondary: secondary,
      },
    });
    console.log(adaptive);
    this.forceUpdate();
  }

  setColor() {
    console.log(imageUrl);
    if (imageUrl !== '') {
      Vibrant.from(imageUrl).getPalette().then((palette) => primary = rgbToHex(Math.round(palette.Vibrant._rgb[0]), Math.round(palette.Vibrant._rgb[1]), Math.round(palette.Vibrant._rgb[2])));
      theme = createMuiTheme( {
        palette: {
          type: themeMode,
          primary: {
            main: primary
          },
          secondary: secondary,
        },
      });
      this.forceUpdate();
    }
  }

  updateTheme() {
    adaptive = true;
    console.log(adaptive);
    this.setColor();
    var self = this
    setTimeout(function() {
      self.setColor();
    }, 250);
    clearTimeout();
  }

  themeModeToggle() {
    if (themeMode === "dark") {
      themeMode = "light";
      theme = createMuiTheme( {
        palette: {
          type: themeMode,
          primary: {
            main: primary
          },
          secondary: secondary,
        },
      });
    }
    else {
      themeMode = "dark";
      theme = createMuiTheme( {
        palette: {
          type: themeMode,
          primary: {
            main: primary
          },
          secondary: secondary,
        },
      });
    }
    this.forceUpdate();
  }

  // to update whenever new song starts playing
  /*
  componentDidUpdate() {
    console.log(timeRemaining);
    if (timeRemaining != 0) {
      const timer = setTimeout(() => {
        this.getNowPlaying();
        if (adaptive == true) {
          this.updateTheme();
        }
      }, timeRemaining);
      return() => clearTimeout(timer);
    }
    else {
      const timer = setTimeout(() => {
        console.log("error");
      }, 5000);
      return() => clearTimeout(timer);
    }
  }
  */

  onKeyDown(keyName, e, handle) {
    if (keyName === 'left') {
      this.getLastSong(); 
    }
    else if (keyName === 'right') {
      this.getNextSong();
    }
    else if (keyName === 'space') {
      this.getPause();
    }
    else if (keyName === 's') {
      this.getShuffle();
    }
    else if (keyName === 'r') {
      this.refresh();
    }
  }

  render() {
    return (
    <div className="App">
      <Hotkeys 
        keyName="right,left,space,s,r" 
        onKeyDown={this.onKeyDown.bind(this)}
      ></Hotkeys>
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <div>
          <NavBar />
        </div>
        <a href='http://localhost:4002/login'> 
        <Button variant="contained" color="primary">Login with Spotify</Button> 
        </a>
        <div>
          <Button variant="contained" color="primary" onClick={() => this.refresh()}>Refresh</Button>
        </div>
        <div>
          <Grid container spacing={3}>
            <Grid item xs={4}>
              <ExpansionPanel>
                <ExpansionPanelSummary expandIcon={<ExpandMoreIcon color='primary'/>}>
                  <Typography>Artist Profile</Typography>
                </ExpansionPanelSummary>
                <ExpansionPanelDetails padding="0">
                <ArtistProfile spotifyApi={spotifyWebApi} onRef={ref => (this.ArtistProfile = ref)}/>
                </ExpansionPanelDetails>
              </ExpansionPanel>
            </Grid>
            <Grid item xs={4}>
              <img src={this.state.nowPlaying.image} alt="album cover" width={300} height={300} mode='fit' style = {{windows: 100}}/>
            </Grid>
            <Grid item xs={4}>
              <ExpansionPanel>
              <ExpansionPanelSummary expandIcon={<ExpandMoreIcon color='primary'/>}>
                <Typography>Lyrics</Typography>
              </ExpansionPanelSummary>
              <ExpansionPanelDetails padding="0">
                <Lyrics spotifyApi={spotifyWebApi} onRef={ref => (this.Lyrics = ref)} />
              </ExpansionPanelDetails>
            </ExpansionPanel>
            </Grid>
          </Grid>
        </div>
        <div><Button variant='outlined'>{this.state.nowPlaying.songInfo}</Button></div>
        <div> 
        <ButtonGroup
          variant="contained"
          color="primary">
            <Button onClick={() => this.getLastSong()}> 
            <Icon>skip_previous</Icon>
          </Button>
          <Button onClick={() => this.getPause()}> 
            {playIcon}
          </Button>
          <Button onClick={() => this.getNextSong()}>
            <Icon>skip_next</Icon>
          </Button>
          <Button onClick={() => this.getShuffle()}>
            {shuffleIcon}
          </Button>
        </ButtonGroup>
        </div>  
        <div>
          <ExpansionPanel>
            <ExpansionPanelSummary
              expandIcon={<ExpandMoreIcon color='primary'/>}
            >
              <Typography>Song Recommendations</Typography>
            </ExpansionPanelSummary>
            <ExpansionPanelDetails>
              <RecommendationsTable/>
            </ExpansionPanelDetails>
          </ExpansionPanel>
        </div>
        <div>
          <ButtonGroup variant='contained' color='primary'>
            <Button onClick={() => this.defaultTheme()}>Default Theme</Button>
            <Button onClick={() => this.updateTheme()}>Adaptive Theme</Button>
          </ButtonGroup>
        </div>
        <div>
          <Button variant='outlined' color='primary' onClick={() => this.themeModeToggle()}>Dark/Light Mode Toggle</Button>
        </div>
      </MuiThemeProvider> 
    </div>   
  );
  }
}
  
export default App;

