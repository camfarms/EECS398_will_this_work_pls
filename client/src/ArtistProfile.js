import React, {Component} from 'react';
import './ArtistProfile.css';


/* This class is a react component containing the artist profile functionality. In order for this component
    to function properly, you must pass it a Spotify object from spotify-web-api-js through props.spotifyApi.
    Assuming a valid and functional Spotify object is passed in, this allows this component to view what song 
    the user is currently playing.
    
    To refresh the artist profile, simply call refreshArtist()*/
class ArtistProfile extends Component {

    //Make sure to pass the spotify api object as a prop
    constructor() {
        super();
        this.state = {
            artist: undefined,
            artistInfo: ""
        }
        this.refreshArtist = this.refreshArtist.bind(this);
    }

    /*Get the artist object of the main artist of the currently playing song.
      Returns a Promise that resolves if this component's state is succesfully updated to the currently
      playing artist.
      Returns a Promise so that you can ensure the currently playing artist is retrieved before doing anything
      else such as looking up that artist's info. */
    async getArtist() {
        var self = this;
        return new Promise(function(resolve, reject) {
            if (!(self.props.spotifyApi === undefined)) {
                self.props.spotifyApi.getMyCurrentPlaybackState().then((response) => {
                    if (!(response.item === undefined)) {
                        self.setState({ 
                            artist: response.item.artists[0].name
                        })
                        resolve();
                    } else {reject()}
                })
            } else {reject()}
        });
    }

    /*Updates this component's state so that it has the currently playing artist's info. */
    getArtistInfo() {
        var self = this;
        var wikiApiUrl = 'https://en.wikipedia.org/w/api.php?format=json&action=query&prop=extracts&exintro' +
            '&explaintext&redirects=1&origin=*&indexpageids&titles=' + self.state.artist;
        fetch(wikiApiUrl)
        .then(response => {
            return response.json();
        }).then(data => {
            var pageid = data.query.pageids[0];
            let info = data['query']['pages'][pageid]['extract'];
            self.setState({artistInfo: info});
        })
    }


    async refreshArtist() {
        await this.getArtist();
        this.getArtistInfo();        
    }

    componentDidMount() {
        this.props.onRef(this);
    }

    componentWillUnmount() {
        this.props.onRef(undefined);
    }

    render() {

        return (
            <div class='artist'>
                {this.state.artistInfo}
            </div>

        )
    }

}

export default ArtistProfile;