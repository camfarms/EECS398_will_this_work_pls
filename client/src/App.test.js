import React from 'react';
import App from './App.js';
import ArtistProfile from './ArtistProfile.js';
import renderer from 'react-test-renderer';
import {render, fireEvent} from '@testing-library/react';

describe('App component tests', () => {

it('renders App component', () => {
    const container = render(<App/>);
    expect(container.firstChild).toMatchSnapshot();
});

});

describe('Artist Profile component tests', () => {

it('renders Artist Profile', () => {
    const container = render(<ArtistProfile spotifyApi={undefined} onRef={ref => (undefined)}/>);
    expect(container.firstChild).toMatchSnapshot();
});

it('Test wikipedia calls', () => {
    const spotifyMock = class {
        getMyCurrentPlaybackState() {
            return new Promise(function(resolve, reject) {
                var response;
                response.item.artists[0].name = 'Wu-tang clan';
                return response;
                resolve();
            });
        }
    }
    const container = render(<ArtistProfile spotifyApi={spotifyMock} onRef={ref => (undefined)}/>);
    
});

});