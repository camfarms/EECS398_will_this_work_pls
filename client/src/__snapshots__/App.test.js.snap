// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> spec renders App component 1`] = `undefined`;

exports[`App component tests renders App component 1`] = `undefined`;

exports[`App renders normally 1`] = `
<div
  className="App"
>
  <div>
    <header>
      <b>
        <font
          size="12"
        >
          Tortoise
        </font>
      </b>
      <img
        height={50}
        mode="fit"
        src="logo.png"
        width={50}
      />
    </header>
  </div>
  <a
    href="http://localhost:4002"
  >
    <button
      className="btn btn-success"
      disabled={false}
      type="button"
    >
      Login with Spotify
    </button>
  </a>
  <div>
     Now Playing: 
    Not Checked
     
  </div>
  <div>
    <img
      src=""
      style={
        Object {
          "widows": 100,
        }
      }
    />
  </div>
  <div>
    <div>
      <h1>
        Artist: 
         
      </h1>
      
    </div>
  </div>
  <div
    className="btn-group"
    role="group"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Check Now Playing
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Get Song Recommendations
    </button>
  </div>
  <div>
    <div />
  </div>
  <div>
    <div
      className="btn-group"
      role="group"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
        Lyrics
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
        Artist Facts
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
        Song Facts
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
         Visuals (Fractals)
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
         Recommended Songs
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
         Settings
      </button>
    </div>
  </div>
</div>
`;

exports[`Artist Profile component tests renders Artist Profile 1`] = `undefined`;

exports[`test 1`] = `
<div
  className="App"
>
  <div>
    <header>
      <b>
        <font
          size="12"
        >
          Tortoise
        </font>
      </b>
      <img
        height={50}
        mode="fit"
        src="logo.png"
        width={50}
      />
    </header>
  </div>
  <a
    href="http://localhost:4002"
  >
    <button
      className="btn btn-success"
      disabled={false}
      type="button"
    >
      Login with Spotify
    </button>
  </a>
  <div>
     Now Playing: 
    Not Checked
     
  </div>
  <div>
    <img
      src=""
      style={
        Object {
          "widows": 100,
        }
      }
    />
  </div>
  <div>
    <div>
      <h1>
        Artist: 
         
      </h1>
      
    </div>
  </div>
  <div
    className="btn-group"
    role="group"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Check Now Playing
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Get Song Recommendations
    </button>
  </div>
  <div>
    <div />
  </div>
  <div>
    <div
      className="btn-group"
      role="group"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
        Lyrics
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
        Artist Facts
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
        Song Facts
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
         Visuals (Fractals)
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
         Recommended Songs
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        type="button"
      >
         Settings
      </button>
    </div>
  </div>
</div>
`;
